core/src/main/java/org/apache/calcite/adapter/enumerable/RexImpTable.java

```diff
@@ -3406,9 +3406,22 @@ abstract Expression implementSafe(RexToLixTranslator translator,
       translatedOperands.add(wmColExpr);
       translatedOperands.add(intervalExpression);
 
-      return Expressions.call(BuiltInMethod.TUMBLING.method, inputEnumerable,
-          EnumUtils.tumblingWindowSelector(inputPhysType, outputPhysType,
-              translatedOperands.get(0), translatedOperands.get(1)));
+      // handle the optional offset parameter. Use 0 for the default value when offset
+      // parameter is not set.
+      final Expression offsetExpr = call.getOperands().size() > 2
+          ? translator.translate(call.getOperands().get(2))
+          : Expressions.constant(0, long.class);
+      translatedOperands.add(offsetExpr);
+
+      return Expressions.call(
+          BuiltInMethod.TUMBLING.method,
+          inputEnumerable,
+          EnumUtils.tumblingWindowSelector(
+              inputPhysType,
+              outputPhysType,
+              translatedOperands.get(0),
```

Comment:
Is there any special reason we must use a list here to keep the translatedOperands ?

Reply:
Ah I think `translatedOperands` might be easy to read. For example, it can be mapped to function signature. When I write such code, I tend to map parameters to offsets. WatermarkCol expression is the first parameter, then others. Maybe not everyone thinks in this way.

I guess you are thinking why not directly use expressions?

Reply:
OK I have updated to remove the list for all three functions.